// @platformatic/kafka/dist/apis/admin/alter-partition-reassignments-v1.js
import { ResponseError } from "../../errors.js";
import { Writer } from "../../protocol/writer.js";
import { createAPI } from "../definitions.js";
/*
  AlterPartitionReassignments Request (Version: 1) => timeout_ms allow_replication_factor_change [topics] TAG_BUFFER
    timeout_ms => INT32
    allow_replication_factor_change => BOOLEAN
    topics => name [partitions] TAG_BUFFER
      name => COMPACT_STRING
      partitions => partition_index [replicas] TAG_BUFFER
        partition_index => INT32
        replicas => INT32
*/
export function createRequest(timeoutMs, allowReplicationFactorChange, topics) {
    return Writer.create()
        .appendInt32(timeoutMs)
        .appendBoolean(allowReplicationFactorChange)
        .appendArray(topics, (w, t) => {
        w.appendString(t.name).appendArray(t.partitions, (w, p) => {
            w.appendInt32(p.partitionIndex).appendArray(p.replicas, (w, r) => w.appendInt32(r), true, false);
        });
    })
        .appendTaggedFields();
}
/*
  AlterPartitionReassignments Response (Version: 1) => throttle_time_ms error_code error_message [responses] TAG_BUFFER
    throttle_time_ms => INT32
    error_code => INT16
    error_message => COMPACT_NULLABLE_STRING
    responses => name [partitions] TAG_BUFFER
      name => COMPACT_STRING
      partitions => partition_index error_code error_message TAG_BUFFER
        partition_index => INT32
        error_code => INT16
        error_message => COMPACT_NULLABLE_STRING
*/
export function parseResponse(_correlationId, apiKey, apiVersion, reader) {
    const errors = [];
    const throttleTimeMs = reader.readInt32();
    const errorCode = reader.readInt16();
    const errorMessage = reader.readNullableString();
    /* c8 ignore next 3 - Hard to test */
    if (errorCode !== 0) {
        errors.push(['', [errorCode, errorMessage]]);
    }
    const response = {
        throttleTimeMs,
        errorCode,
        errorMessage,
        responses: reader.readArray((r, i) => {
            return {
                name: r.readString(),
                partitions: r.readArray((r, j) => {
                    const partition = {
                        partitionIndex: r.readInt32(),
                        errorCode: r.readInt16(),
                        errorMessage: r.readNullableString()
                    };
                    if (partition.errorCode !== 0) {
                        errors.push([`responses/${i}/partitions/${j}`, [partition.errorCode, partition.errorMessage]]);
                    }
                    return partition;
                })
            };
        })
    };
    if (errors.length) {
        throw new ResponseError(apiKey, apiVersion, Object.fromEntries(errors), response);
    }
    return response;
}
export const api = createAPI(45, 1, createRequest, parseResponse);
